File: 101.symmetric-tree.java
/*
 * @lc app=leetcode id=101 lang=java
 *
 * [101] Symmetric Tree
 */

// @lc code=start

import java.util.ArrayDeque;
import java.util.LinkedList;
import java.util.Queue;

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 * int val;
 * TreeNode left;
 * TreeNode right;
 * TreeNode() {}
 * TreeNode(int val) { this.val = val; }
 * TreeNode(int val, TreeNode left, TreeNode right) {
 * this.val = val;
 * this.left = left;
 * this.right = right;
 * }
 * }
 */
class Solution {
    public boolean isSymmetric(TreeNode root) {
        if (root == null)
            return true;
        Queue<TreeNode> queue = new LinkedList<>();
        queue.add(root.left);
        queue.add(root.right);
        while (!queue.isEmpty()) {
            TreeNode lnode = queue.remove();
            TreeNode rnode = queue.remove();
            if (lnode == null && rnode == null)
                continue;
            if (lnode == null ^ rnode == null)
                return false;
            if (lnode.val != rnode.val)
                return false;
            queue.add(lnode.left);
            queue.add(rnode.right);
            queue.add(lnode.right);
            queue.add(rnode.left);
        }
        return true;
    }
}
// @lc code=end

File: 101.symmetric-tree.cpp
/*
 * @lc app=leetcode id=101 lang=cpp
 *
 * [101] Symmetric Tree
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

#include <queue>

using namespace std;

class Solution {
public:
    bool isSymmetric(TreeNode *root) {
        if (root == nullptr) return true;
        queue<TreeNode *> q;
        q.push(root->left); q.push(root->right);
        while (!q.empty()) {
            TreeNode *left = q.front(); q.pop();
            TreeNode *right = q.front(); q.pop();
            if (left == nullptr && right == nullptr) continue;
            if (left == nullptr ^ right == nullptr) return false;
            if (left->val != right->val) return false;
            q.push(left->left); q.push(right->right);
            q.push(left->right); q.push(right->left);
        }
        return true;
    }
};
// @lc code=end
